src/Orleans.Core/Async/TaskExtensions.cs

```diff
@@ -258,6 +258,90 @@ public static async Task<T> WithTimeout<T>(this Task<T> taskToComplete, TimeSpan
             throw new TimeoutException(errorMessage);
         }
 
+        /// <summary>
+        /// For making an uncancellable task cancellable, by ignoring its result.
+        /// </summary>
+        /// <param name="taskToComplete">The task to wait for unless cancelled</param>
+        /// <param name="cancellationToken">A cancellation token for cancelling the wait</param>
+        /// <param name="message">Message to set in the exception</param>
+        /// <returns></returns>
+        internal static async Task WithCancellation(
+            this Task taskToComplete,
+            CancellationToken cancellationToken,
+            string message)
+        {
+            try
+            {
+                await taskToComplete.WithCancellation(cancellationToken);
+            }
+            catch (TaskCanceledException ex)
+            {
+                throw new TaskCanceledException(message, ex);
+            }
+        }
+
+        /// <summary>
+        /// For making an uncancellable task cancellable, by ignoring its result.
+        /// </summary>
+        /// <param name="taskToComplete">The task to wait for unless cancelled</param>
+        /// <param name="cancellationToken">A cancellation token for cancelling the wait</param>
+        /// <returns></returns>
+        internal static Task WithCancellation(this Task taskToComplete, CancellationToken cancellationToken)
+        {
+            if (taskToComplete.IsCompleted || !cancellationToken.CanBeCanceled)
+            {
+                return taskToComplete;
+            }
+            else if (cancellationToken.IsCancellationRequested)
+            {
+                return TaskFromCanceled<object>();
```

Comment:
I think we can use `Task.FromCanceled<object>()` instead and eventually deprecate `TaskFromCanceled<T>`